api/http-request-practice.md

```diff
@@ -1,90 +1,261 @@
+- Once the promise resolves, response will contain a JavaScript object representing the response from the server
+- To see the body as like above, execute `await response.json()`
+
+What happens?
 
-- Example
-- Example
+- [b0rk explains CORS](https://twitter.com/b0rk/status/1162392625057583104)
+- [Hacking It Out: When CORS won’t let you be great](https://medium.com/netscape/hacking-it-out-when-cors-wont-let-you-be-great-35f6206cc646)
 
+To keep things simple, for this example, we just load the same origin, worldtimeapi.org, so we didn't have to worry about it.
 
 ### Guided Practice
 
-Have the apprentices work with you as you do something step-by-step.  This can also be fulfilled by a detailed tutorial intended for beginners.
+`fetch` is a standard added by browsers but if you like it, it's also available as a [package for node](https://www.npmjs.com/package/node-fetch). Let's use it!
+
+- `npm install --save node-fetch`
+- open node REPL
+- `require('node-fetch')`
+- Run the same command from fetch in the browser but in your node repl
 
 
 ### Independent Practice
 
-Class does this thing themselves with specific additional items. This could be alone, with a partner, or small group; but the idea is that it's less guided, more independent.
+Since there are tons of libraries out there, now's a great time to try a few and see what you like/dislike about them compared to those we used above.
+
+- [5 Ways to Make HTTP Requests in Node.js](https://www.twilio.com/blog/2017/08/http-requests-in-node-js.html)
+
+*Exercise:* 
+
+- Choose 2 libraries from the above article and take a few notes comparing them.
+- Which was your most favorite of the 5 + 2? Your _least favorite_?
+
 
 ### Challenge
 
-Apprentices can try to do this other thing. Ideally, they will be challenged to connect what they've learned to some previous knowledge or additional research.
+#### Advanced: Node.js Web Server
+You have been using Express to implement a web server. Node also has a lower level library, which is part of the same library as `http.request`. It is called `http.server`.
+
+Find a tutorial to make the simplest possible Node web server using _only_ `http.server` and make some requests against it with your favorite HTTP client library from above.
+
+#### Advanced: DIY using telnet
+
+- `brew install telnet` if not installed
+- format is `telnet server port` (note: it's not a colon like in a URL)
+- `telnet www.google.com 80`
+- It will connect and inform you of the escape character
+- type `GET / HTTP/1.1` then enter twice. This is the smallest valid HTTP request
+- A Google web server will spit out the Google home page
 
+If you're curious what each required piece is, check out this [simplified HTTP specification](https://www.jmarshall.com/easy/http/#requestline). The [complete spec](https://www.w3.org/Protocols/rfc2616/rfc2616.html) is huge and uses a formal mathematical language if you want to go down a rabbit hole.
+
+#### Discussion Question
```

Comment:
Can you combine this section down with the favorite tools question below?

Reply:
moved

Reply:
Can you move all of the discussion questions into CFU, I mean?